Aug 4 2021/Ch1ResnetV3.py

```diff
@@ -1,245 +1,225 @@
+        # targets = labels
+        loss = train_batch(data, targets, model, optimizer, criterion) # training loss of our model
+        example_ct += len(data) # I do not understand what this is doing
+        batch_ct += 1 # I also do not understand what this is doing
+        loss_list.append(loss.item()) # add our loss defined in line 117 to our  empty loss list defined above in line 45
+        if((batch_ct + 1 ) % 5) == 0: # report metrics every 5th batch and log the loss for weights and biases
+            train_log(loss, example_ct, hyperparameter.num_epochs)
+            #break
 
-"""
-Step 8 Define Training logic 
-"""
-"""
+def train_batch(data, targets, model, optimizer, criterion): # function that defines the batches for training
+    data, targets = data.to(device) , targets.to(device)
 
-1) track training loss per sample
-    - In trainging loop  wandb.lop({'train_per_sample_loss' : train_loss})
-    
-"""
-def train(model, train_loader, criterion, optimizer, config):
-    # tell wandb to watch what the model gets ups to: gradients, weights and more
-    # Run training and track with wandb
-    total_batches = len(train_loader) * config.num_epochs
-    example_ct = 0
-    batch_ct = 0
-
-    for batch_idx, (data, targets) in enumerate(train_loader):
-        loss = train_batch(data, targets, model, optimizer, criterion)
-        example_ct += len(data)
-        batch_ct += 1
-        # report metrics every 5th batch
-        loss_list.append(loss.item())
-        if((batch_ct + 1 ) % 5) == 0:
-            train_log(loss, example_ct, config.num_epochs)
-            #break
 
-def train_batch(data, targets, model, optimizer, criterion):
-    data, targets = data.to(device), targets.to(device)
+    # data = data.to(device) # send the images to the device either GPU or CPU
+    # targets = targets.to(device) # Send the labels to the device either GPU or CPU
 
     # forward pass
-    scores = model(data)
-    loss = criterion(scores, targets)
-    # outputs = model(img)
+    scores = model(data) # scores = the model trained on the images
+    loss = criterion(scores, targets) # loss = MSE loss evaluated with the scores and the labels
 
-    # loss = criterion(outputs,y_label)
     # backward pass
-    optimizer.zero_grad()
+    optimizer.zero_grad() # Pass the optimizer through the model
     loss.backward()
     # step with optimizer
     optimizer.step()
-    return loss
-def train_log(loss, example_ct, epoch):
-    train_loss = float(loss)
 
+    return loss # return the loss
+def train_log(loss, example_ct, epoch): # define what we want to log with weights and biases
+    train_loss = float(loss) # convert the loss from a torch float tensor to a float
+    wandb.log({'train_per_sample_loss': train_loss}) # log the train per sample loss
+    wandb.log({'train_average_loss': np.mean(loss_list)}) # calculate and log the average loss
```

Comment:
see above comment, I don't think you want to log the average loss every timestep, this should only be called after the final iteration of each epoch, aka line 145 after the training for loop is done. You did it correctly for the Validation_average_loss (i.e., outside the for loop)